### PR TITLE
TandemFoil Paper: weight decay sweep (WD=5e-3 and WD=2e-2)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The TFP champion uses WD=1e-2 (transferred from the TF champion config). Weight decay is an untested knob on TFP — the value was inherited rather than optimized.

- **WD=5e-3 (halved):** Less regularization may allow the model to fit the field_mse targets more closely, especially the pressure channel which requires precise predictions in asinh-transformed space.
- **WD=2e-2 (doubled):** Stronger regularization may improve generalization, particularly if the model is overfitting at ep443 on the small TFP validation split.

Since TFP has proven extremely sensitive to hyperparameter changes (T_max, gc, LR, depth all cause Infinity), WD is a softer knob — it doesn't change gradient dynamics as aggressively as LR or gc.

## Instructions

Run two WD values at the exact champion config:

**Run 1: WD=5e-3**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 5e-3 \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --ema-decay 0.999 \
  --wandb_group tfp-wd-sweep
```

**Run 2: WD=2e-2**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 2e-2 \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --ema-decay 0.999 \
  --wandb_group tfp-wd-sweep
```

Report best val_primary/field_mse for both. Also note divergence epoch if applicable — the pressure channel Infinity issue has been the main failure mode on TFP.

## Baseline

- **val_primary/field_mse:** 0.002383 at epoch 443 (diverged ep462)
- **Config:** Lion lr=1.25e-4, T_max=10, gc=0.5, **WD=1e-2**, EMA=0.999, 3L/192d
- **W&B run:** d1xh0o1p
- **Reproduce:** `cd target/icml2026 && python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shinobu:

## Results

### Run 1: WD=5e-3 (halved) — NEGATIVE

| Metric | Baseline (WD=1e-2) | WD=5e-3 | Delta |
|--------|-------------------|---------|-------|
| val_primary/field_mse | 0.002383 (ep443) | **Infinity** (never resolved) | N/A |
| Best epoch | 443 | N/A (pressure never finite) | — |
| Final epoch | 462 (diverged) | 544 (slow divergence) | — |

- **W&B run:** [2h4i8ulf](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/2h4i8ulf)
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 5e-3 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 --agent shinobu --wandb-name "shinobu/tfp-wd-5e-3" --wandb-group tfp-wd-sweep`
- **544 epochs** in 360-min budget (~0.66 min/epoch)
- **What happened:** The pressure channel (surface_mse_p, volume_mse_p) never resolved to finite values during the entire 544-epoch run. The model showed healthy loss descent until epoch ~134 (loss=0.045), then loss monotonically increased for the remaining 410 epochs, reaching 0.249 by ep544. Grad norms climbed steadily from ~14 to ~48. The non-pressure channels (Ux, Uy) improved normally, but the asinh-transformed pressure prediction never stabilized. This is a fundamentally different failure mode from the baseline — the baseline resolved pressure and peaked at ep443, while WD=5e-3 prevented pressure resolution entirely.

### Run 2: WD=2e-2 (doubled) — NEGATIVE

| Metric | Baseline (WD=1e-2) | WD=2e-2 | Delta |
|--------|-------------------|---------|-------|
| val_primary/field_mse | 0.002383 (ep443) | 0.901 (ep242, during divergence) | +37,700% |
| Best epoch | 443 | 242 (divergence epoch) | — |
| Final epoch | 462 (diverged) | 545 (diverged ep240, NaN grads) | — |

- **W&B run:** [rbivtl6r](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/rbivtl6r)
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 2e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 --agent shinobu --wandb-name "shinobu/tfp-wd-2e-2" --wandb-group tfp-wd-sweep`
- **545 epochs** in 360-min budget (~0.66 min/epoch)
- **What happened:** Similar to WD=5e-3, the pressure channel never resolved before the model diverged. Catastrophic divergence hit at epoch 240 — loss exploded from ~0.054 to 3.3→9.0→20.9 over 3 epochs, grad norms went to Infinity then NaN. The only finite field_mse (0.901) appeared during the divergence epoch itself and is not meaningful. The model continued running post-divergence for 305 more epochs but produced only garbage. Stronger regularization slowed pressure resolution even further than WD=5e-3 and triggered earlier catastrophic instability.

## Analysis

**Both WD variants are strictly worse than the baseline WD=1e-2.** The key finding is that weight decay is not a "soft knob" on TFP — it critically interacts with the pressure channel resolution mechanism in asinh-transformed space:

1. **WD=1e-2 (baseline):** Pressure resolves, field_mse reaches 0.002383 at ep443, diverges at ep462.
2. **WD=5e-3 (halved):** Pressure **never resolves** in 544 epochs. Slow divergence from ep134. Less regularization paradoxically prevented pressure convergence.
3. **WD=2e-2 (doubled):** Pressure **never resolves** before catastrophic divergence at ep240. More regularization slowed pressure resolution and brought forward the T_max=10 instability.

The fact that neither direction worked suggests WD=1e-2 is at or near a critical sweet spot for the TFP pressure channel — too little WD and the model fails to learn the pressure mapping, too much and it diverges before it can.

## Suggested follow-ups

- **WD=8e-3 and WD=1.2e-2** — very narrow sweep around the working value to test the sweet-spot hypothesis
- **Investigate why WD interacts with pressure resolution** — the non-pressure channels (Ux, Uy) were healthy on both runs. The failure is specific to the asinh-pressure prediction pathway
- **T_max=20 or T_max=30 with WD variants** — longer cosine periods might give the pressure channel more time to resolve before instability hits

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3133 because ADVISOR: Closing — WD=5e-3 produced Infinity (pressure channel diverged). WD=1e-2 confirmed as the correct value for TFP. Shinobu reassigned..
